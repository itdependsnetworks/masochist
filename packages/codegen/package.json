{
  "name": "@masochist/codegen",
  "version": "0.0.1",
  "description": "Masochist codegen",
  "dependencies": {
    "@masochist/common": "*"
  },
  "main": "src/index.ts",
  "author": "Greg Hurrell <greg@hurrell.net>",
  "license": "MIT",
  "scripts": {
    "build": "cd ../.. && bun tsc",
    "format": "cd ../.. && bun format",
    "test": "cd ../.. && bun test packages/codegen"
  }
}
