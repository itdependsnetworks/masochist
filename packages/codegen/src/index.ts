export {default as ast} from './ast';
export {default as print} from './print';
export {default as quote} from './quote';
export {default as walk} from './walk';

export type {
  Argument,
  ArrayValue,
  AssignmentStatement,
  BinaryExpression,
  BooleanValue,
  CallExpression,
  ClassDeclaration,
  Consequent,
  Declaration,
  DocComment,
  ExportDefaultDeclaration,
  ExportNamedDeclaration,
  Expression,
  ExpressionStatement,
  FunctionDeclaration,
  FunctionExpression,
  GetAccessor,
  Identifier,
  IfStatement,
  ImportStatement,
  MemberExpression,
  MethodDefinition,
  NewExpression,
  Node,
  NullValue,
  NumberValue,
  ObjectProperty,
  ObjectValue,
  PrimitiveValue,
  Program,
  PropertyDeclaration,
  RawStatement,
  ReturnStatement,
  Statement,
  StringValue,
  SwitchCase,
  SwitchStatement,
  WhileStatement,
} from './ast';
