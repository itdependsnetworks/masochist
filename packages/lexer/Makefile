.DEFAULT: png

DOT = $(wildcard diagrams/*.dot)
PNG = $(patsubst %.dot,%.png,$(DOT))

.PHONY: clean
clean:
	rm -f diagrams/*

.PHONY: debug
debug:
	@echo DOT
	@echo $(DOT)
	@echo PNG
	@echo $(PNG)

.PHONY: diagrams
diagrams: $(PNG)

%.png: %.dot
	dot -Tpng $< -o $@
