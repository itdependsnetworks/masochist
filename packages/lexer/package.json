{
  "name": "@masochist/lexer",
  "version": "0.0.1",
  "description": "Masochist lexer",
  "dependencies": {
    "@masochist/codegen": "*",
    "@masochist/common": "*"
  },
  "main": "lib/index.js",
  "author": "Greg Hurrell <greg@hurrell.net>",
  "license": "MIT",
  "scripts": {
    "build": "cd ../.. && yarn build",
    "format": "cd ../.. && yarn format",
    "test": "cd ../.. && yarn test packages/lexer"
  }
}
