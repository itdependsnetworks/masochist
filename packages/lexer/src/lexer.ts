import generate from './generate';

//
// Punctuators (2.1.8)
//
export const AMPERSAND = '&';
export const AT = '@';
export const BANG = '!';
export const BAR = '|';
export const CLOSING_BRACE = '}';
export const CLOSING_BRACKET = ']';
export const CLOSING_PAREN = ')';
export const COLON = ':';
export const DOLLAR = '$';
export const ELLIPSIS = '...';
export const EQUALS = '=';
export const OPENING_BRACE = '{';
export const OPENING_BRACKET = '[';
export const OPENING_PAREN = '(';

export const ESCAPED_CHARACTER = /\\["\\\/bfnrt]/;

// NOTE: Don't use "i" flag with this RegExp, because "\U" is not valid.
export const ESCAPED_UNICODE = /\\u[0-9A-Fa-f]{4}/;

export const COMMA = /,/;

// Allowed characters = SOURCE_CHARACTER - LINE_TERMINATOR ([\u000a\u000d]).
export const COMMENT = /#[\u0009\u0020-\uffff]*/;

export const EXPONENT_PART = /[eE][+-]?\d+/;
export const FRACTIONAL_PART = /\.\d+/;
export const INTEGER_PART = /-?(0|[1-9]\d*)/;
export const LINE_TERMINATOR = /\n|\r\n|\r/;
export const NAME = /[_a-z][_0-9a-z]*/i;
export const SOURCE_CHARACTER = /[\u0009\u000a\u000d\u0020-\uffff]/;
export const UNICODE_BOM = /\ufeff/;
export const WHITESPACE = /[\t ]+/;

// This rather ghastly RegExp is made up of any of:
//
// - ESCAPED_UNICODE
// - ESCAPED_CHARACTER
// - SOURCE_CHARACTER, minus \u000a (\n), \u000d (\r), \u0022 ("), and \u005c (\)
//
export const STRING_CHARACTER = /(\\u[0-9A-Fa-f]{4}|\\["\\\/bfnrt]|[\u0009\u0020\u0021\u0023-\u005b\u005d-\uffff])/;

/**
 * Generate a lexer for the GraphQL language.
 *
 * GraphQL is a simple language where almost all tokens can be recognized with a
 * single character of lookahead. The only exceptions are:
 *
 * - BLOCK_STRING_VALUE and STRING_VALUE: These require three characters of
 *   lookahead, to distinguish between the former (delimited by '"""') and the
 *   latter (delimited by '"').
 *
 *   A STRING_VALUE may contain escapes, but a BLOCK_STRING_VALUE is effectively
 *   "raw", other than some special whitespace handling dictated by the spec:
 *   - https://spec.graphql.org/October2021/#sec-String-Value.Block-Strings
 *
 * - FLOAT_VALUE and INT_VALUE: An INT_VALUE may not be followed by a dot or a
 *   NAME. As an added wrinkle, an INT_VALUE of zero may not be followed by
 *   another digit (ie. "00" is invalid, as is "01" etc). If an INT_VALUE is
 *   followed by a dot or "e"/"E" then it must be interpreted as a FLOAT_VALUE
 *   instead.
 *
 *   A FLOAT_VALUE must not be followed by a dot (ie. "1.2.3" is invalid) or a
 *   NAME (ie. "1.2L" is invalid).
 *
 *   See:
 *   - https://spec.graphql.org/October2021/#sec-Int-Value
 *   - https://spec.graphql.org/October2021/#sec-Float-Value
 */
export default generate(({ignored, token}) => {
  //
  // Punctuators (2.1.8)
  //
  token('BANG', '!');
  token('DOLLAR', '$');
  token('AMPERSAND', '&');
  token('OPENING_PAREN', '(');
  token('CLOSING_PAREN', ')');
  token('ELLIPSIS', '...');
  token('COLON', ':');
  token('EQUALS', '=');
  token('AT', '@');
  token('OPENING_BRACKET', '[');
  token('CLOSING_BRACKET', ']');
  token('OPENING_BRACE', '{');
  token('BAR', '|');
  token('CLOSING_BRACE', '}');

  //
  // (Other) lexical tokens (2.1.6).
  //
  token('NAME', NAME);

  // TODO: figure out how to do this without back-tracking
  // about numbers: see recent changes to spec:
  // https://github.com/graphql/graphql-spec/pull/601
  //
  // FLOAT_VALUE:
  //
  //    const FLOAT_VALUE = sequence(
  //      'INTEGER_PART',
  //      oneOf(
  //        sequence('FRACTIONAL_PART', 'EXPONENT_PART'),
  //        'EXPONENT_PART',
  //        'FRACTIONAL_PART',
  //      ),
  //    );
  //
  // where the various `*_PART` patterns are defined as follows:
  //
  //    match(/[eE][+-]?\d+/).name('EXPONENT_PART');
  //    match(/\.\d+/).name('FRACTIONAL_PART');
  //    match(/-?(?:0|[1-9]\d*)/).name('INTEGER_PART');
  //
  // INT_VALUE:
  //
  //    const INT_VALUE = an('INTEGER_PART');

  token(
    'BLOCK_STRING_VALUE',
    '"""',
    // TODO: BLOCK_STRING_CHARACTER
    '"""',
  );
  // cf previous lexer:
  //
  //    const BLOCK_STRING_VALUE = sequence(
  //      match('"""'),
  //      maybe(repeat('BLOCK_STRING_CHARACTER')),
  //      match('"""'),
  //    );
  //
  // where BLOCK_STRING_CHARACTER is:
  //
  //    oneOf(
  //      match('\\"""'),
  //      a('SOURCE_CHARACTER').except(oneOf(match('"""'), match('\\"""'))),
  //    ).name('BLOCK_STRING_CHARACTER');

  token(
    'STRING_VALUE',
    '"',
    STRING_CHARACTER,
    '"',
  );
  // cf previous lexer:
  //
  //    const STRING_VALUE = sequence(
  //      match('"'),
  //      maybe(repeat('STRING_CHARACTER')),
  //      match('"'),
  //    );
  //
  // where STRING_CHARACTER is:
  //
  //    oneOf(
  //      a('SOURCE_CHARACTER').except(
  //        oneOf(match('"'), match('\\'), 'LINE_TERMINATOR'),
  //      ),
  //      'ESCAPED_UNICODE',
  //      'ESCAPED_CHARACTER',
  //    ).name('STRING_CHARACTER');

  //
  // Ignored tokens (2.1.7).
  //
  ignored('UNICODE_BOM', UNICODE_BOM);
  ignored('WHITESPACE', WHITESPACE);
  ignored('LINE_TERMINATOR', LINE_TERMINATOR);
  ignored('COMMENT', COMMENT);
  ignored('COMMA', COMMA);
});

// Expect generated lexer to look something like the following...
//
// eg. for NAME translation table:
//
//    {
//      acceptStates: new Set([1]),
//      startStates: new Set([0]),
//      transitions: [
//        /* 0 */ new Map([
//          ['Range:a-z', new Set([1])],
//          ['Range:A-Z', new Set([1])],
//          ['Atom:_', new Set([1])],
//        ]),
//        /* 1 */ new Map([
//          ['Range:a-z', new Set([1])],
//          ['Range:A-Z', new Set([1])],
//          ['Atom:_', new Set([1])],
//          ['Range:0-9', new Set([1])],
//        ]),
//      ],
//    }

// TODO: add invariant; as these are DFAs, the target Set should only ever have
// one member (a DFA is a special case of NFA, but we are only showing NFA-ness
// in the typesystem right now and to be honest I am not sure we _could_ do
// anything else; we can in the transition table at least)

const input = 'something';
let i = 0;
const start = 0;
let state = 0;
let accept = null;

const ACCEPT = -2;
const REJECT = -1;

main_loop: while (i < input.length) {
  const ch = input[i];
  switch (state) {
    case 0:
      if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch === '_') {
        state = 1;
      } else {
        state = REJECT;
      }
      break;
    case 1:
      if (
        (ch >= 'a' && ch <= 'z') ||
        (ch >= 'A' && ch <= 'Z') ||
        ch === '_' ||
        (ch >= '9' && ch <= '9')
      ) {
        state = 1;
      } else {
        state = ACCEPT;
      }
      break;
    case REJECT:
      // We shouldn't even be trying to recognize a token unless lookahead tells
      // us to, so failure means invalid input.
      throw new Error('failed to recognize token');
    case ACCEPT:
      accept = 'NAME';
      break main_loop;
  }
  i++;
}

// Probably want Ragel-like actions that I can define to emit tokens.  And
// specify priority on transitions so I can take the union of machines and have
// them do the right thing faced with "foo \" bar" and such (might even obviate
// the need for lookahead).
