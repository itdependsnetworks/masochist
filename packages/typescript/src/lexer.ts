import {ignore, union} from '@masochist/lexer';

// Operators.
export const ASSIGN = '=';
export const BANG = '!';
export const CLOSING_BRACE = '}';
export const CLOSING_BRACKET = ']';
export const CLOSING_PAREN = ')';
export const COMMA = ',';
export const COLON = ':';
export const DOT = '.';
export const EQUALS = '==';
export const GREATER_THAN = '>';
export const GREATER_THAN_OR_EQUAL = '>=';
export const LESS_THAN = '<';
export const LESS_THAN_OR_EQUAL = '<=';
export const LOGICAL_AND = '&&';
export const LOGICAL_OR = '||';
export const MINUS = '-';
export const OPENING_BRACE = '{';
export const OPENING_BRACKET = '[';
export const OPENING_PAREN = '(';
export const PLUS = '+';
export const REST = '...';
export const SEMICOLON = ';';
export const STRICT_EQUALS = '===';

// Reserved words.
//
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#reserved_words
export const CLASS = 'class';
export const CONST = 'const';
export const CONTINUE = 'continue';
export const DEFAULT = 'default';
export const ELSE = 'else';
export const EXPORT = 'export';
export const FALSE = 'false';
export const FOR = 'for';
export const FUNCTION = 'function';
export const IF = 'if';
export const IMPORT = 'import';
export const LET = 'let';
export const NEW = 'new';
export const NULL = 'null';
export const RETURN = 'return';
export const THIS = 'this';
export const TRUE = 'true';
export const WHILE = 'while';
export const YIELD = 'yield';

// Special identifiers.
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#identifiers_with_special_meanings

export const GET = 'get';
// others: arguments, as, async, eval, from, get, of, set

// TODO: add $ here
// don't bother with non-ASCII identifiers as described in:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#identifiers
export const IDENTIFIER = /[_a-z][_0-9a-z]*/i;
export const NUMBER = /-?(0|[1-9]\d*)(\.\d+)?/;

// String literals.
// See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types

const OCTAL_ESCAPE = /\\[0-7][0-7][0-7]/;
const HEX_ESCAPE = /\\x[0-9A-Fa-f][0-9A-Fa-f]/;
const UNICODE_ESCAPE = /\\u[0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f]/;
const EXTENDED_UNICODE_ESCAPE =
  /\\u\{[0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f][0-9A-Fa-f]\}/;
const CHARACTER_ESCAPE = /\\./;

const SINGLE_QUOTED_STRING = "'(" +
  [
    OCTAL_ESCAPE,
    HEX_ESCAPE,
    UNICODE_ESCAPE,
    EXTENDED_UNICODE_ESCAPE,
    CHARACTER_ESCAPE,
    /[^'\\]/,
  ]
    .map((regExp) => regExp.toString().slice(1, -1))
    .join('|') +
  ")*'";
const DOUBLE_QUOTED_STRING = '"(' +
  [
    OCTAL_ESCAPE,
    HEX_ESCAPE,
    UNICODE_ESCAPE,
    EXTENDED_UNICODE_ESCAPE,
    CHARACTER_ESCAPE,
    /[^"\\]/,
  ]
    .map((regExp) => regExp.toString().slice(1, -1))
    .join('|') +
  ')*"';

export const STRING_VALUE = new RegExp(
  SINGLE_QUOTED_STRING + '|' + DOUBLE_QUOTED_STRING,
);

// Ignored.
export const COMMENT = /\/\/[^\n\r]*(?:\r?\n)?/;
export const LINE_TERMINATOR = /\n|\r\n|\r/; // Note: we require semicolons.
export const WHITESPACE = /[\t ]+/;

export default union({
  // Order matters; reserved keywords and special identifiers must come before
  // normal identifiers.
  CLASS,
  CONST,
  CONTINUE,
  DEFAULT,
  ELSE,
  EXPORT,
  FALSE,
  FOR,
  FUNCTION,
  GET,
  IF,
  IMPORT,
  LET,
  NEW,
  NULL,
  RETURN,
  THIS,
  TRUE,
  WHILE,
  YIELD,

  ASSIGN,
  BANG,
  CLOSING_BRACE,
  CLOSING_BRACKET,
  CLOSING_PAREN,
  COLON,
  COMMA,
  DOT,
  EQUALS,
  GREATER_THAN,
  GREATER_THAN_OR_EQUAL,
  IDENTIFIER,
  LESS_THAN,
  LESS_THAN_OR_EQUAL,
  LOGICAL_AND,
  LOGICAL_OR,
  MINUS,
  NUMBER,
  OPENING_BRACE,
  OPENING_BRACKET,
  OPENING_PAREN,
  PLUS,
  REST,
  SEMICOLON,
  STRICT_EQUALS,
  STRING_VALUE,

  IGNORED: ignore(COMMENT, LINE_TERMINATOR, WHITESPACE),
});
