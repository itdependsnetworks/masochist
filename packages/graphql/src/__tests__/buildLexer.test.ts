// These tests used to live in `@masochist/lexer`, where `build` is defined, but
// they depend on artifacts produced by `@masochist/graphql`, so we've moved
// them in here to avoid a circular dependency.

import {print} from '@masochist/codegen';
import {build} from '@masochist/lexer';
import {describe, expect, it} from 'bun:test';

import lexer from '../lexer';

import type {Stats} from '@masochist/lexer';

describe('build()', () => {
  it('builds an AST', async () => {
    const ast = await build(lexer);
    expect(print(ast)).toMatchSnapshot();
  });

  it('reports stats', async () => {
    const stats: Stats = {};

    await build(lexer, stats);

    expect(stats['startStates']).toBe(1);
    expect(stats['acceptStates']).toBe(101);
    expect(stats['totalStates']).toBe(122);
    expect(stats['ignoredTokens']).toBe(4);
    expect(stats['totalTokens']).toBe(101);
    expect(stats['inlineableAcceptStates']).toBe(16);
    expect(stats['inlinedAcceptStates']).toBe(16);
    expect(stats['inlinedSelfTransitions']).toBe(10);
  });
});
