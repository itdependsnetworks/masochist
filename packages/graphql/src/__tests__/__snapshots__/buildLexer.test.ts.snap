// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`build() builds an AST 1`] = `
"/**
 * vim: set nomodifiable : DO NOT EDIT - edit "build.ts" instead
 *
 * @generated
 */
import {Token} from '@masochist/lexer';
const REJECT = -1;
const START = 0;
export class Lexer {
  input: string;
  state: number;
  tokenStart: number;
  index: number;

  /**
   * @param {string} input
   */
  constructor(input: string) {
    this.input = input;
    this.state = START;
    this.tokenStart = 0;
    this.index = 0;
  }

  /**
   * @param {string} name
   * @param {number} end
   * @param {string} input
   */
  emit(name: string, end: number, input: string) {
    const token = new Token(name, this.tokenStart, end, input);
    this.tokenStart = end;
    this.index = end;
    return token;
  }

  next() {
    const input = this.input;
    const length = input.length;
    while (this.index <= length) {
      const state = this.state;
      let ch = this.index < length ? input.charCodeAt(this.index) : -1;
      if (state === START) {
        if (ch === 0x09 || ch === 0x20) {
          this.state = 1;
        } else if (ch === 0x0a || ch === 0x2c || ch === 0xfeff) {
          this.state = 2;
        } else if (ch === 0x0d) {
          this.state = 3;
        } else if (ch === 0x21) {
          return this.emit('BANG', this.index + 1, input);
        } else if (ch === 0x22) {
          this.state = 5;
        } else if (ch === 0x23) {
          this.state = 6;
        } else if (ch === 0x24) {
          return this.emit('DOLLAR', this.index + 1, input);
        } else if (ch === 0x26) {
          return this.emit('AMPERSAND', this.index + 1, input);
        } else if (ch === 0x28) {
          return this.emit('OPENING_PAREN', this.index + 1, input);
        } else if (ch === 0x29) {
          return this.emit('CLOSING_PAREN', this.index + 1, input);
        } else if (ch === 0x2d) {
          this.state = 11;
        } else if (ch === 0x2e) {
          this.state = 12;
        } else if (ch === 0x30) {
          this.state = 13;
        } else if (ch >= 0x31 && ch <= 0x39) {
          this.state = 14;
        } else if (ch === 0x3a) {
          return this.emit('COLON', this.index + 1, input);
        } else if (ch === 0x3d) {
          return this.emit('EQUALS', this.index + 1, input);
        } else if (ch === 0x40) {
          return this.emit('AT', this.index + 1, input);
        } else if (ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x65 || ch >= 0x67 && ch <= 0x6e || ch >= 0x70 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x5b) {
          return this.emit('OPENING_BRACKET', this.index + 1, input);
        } else if (ch === 0x5d) {
          return this.emit('CLOSING_BRACKET', this.index + 1, input);
        } else if (ch === 0x66) {
          this.state = 21;
        } else if (ch === 0x6f) {
          this.state = 22;
        } else if (ch === 0x7b) {
          return this.emit('OPENING_BRACE', this.index + 1, input);
        } else if (ch === 0x7c) {
          return this.emit('BAR', this.index + 1, input);
        } else if (ch === 0x7d) {
          return this.emit('CLOSING_BRACE', this.index + 1, input);
        } else {
          this.state = REJECT;
        }
      } else if (state === 1) {
        while (ch === 0x09 || ch === 0x20) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        // IGNORED token.
        this.tokenStart = this.index;
        this.state = START;
        continue;
      } else if (state === 2) {
        // IGNORED token.
        this.tokenStart = this.index;
        this.state = START;
        continue;
      } else if (state === 3) {
        if (ch === 0x0a) {
          this.state = 2;
        } else {
          // IGNORED token.
          this.tokenStart = this.index;
          this.state = START;
          continue;
        }
      } else if (state === 5) {
        if (ch === 0x09 || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 26;
        } else if (ch === 0x22) {
          this.state = 27;
        } else if (ch === 0x5c) {
          this.state = 28;
        } else {
          this.state = REJECT;
        }
      } else if (state === 6) {
        while (ch === 0x09 || ch >= 0x20 && ch <= 0xffff) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        // IGNORED token.
        this.tokenStart = this.index;
        this.state = START;
        continue;
      } else if (state === 11) {
        if (ch === 0x30) {
          this.state = 13;
        } else if (ch >= 0x31 && ch <= 0x39) {
          this.state = 14;
        } else {
          this.state = REJECT;
        }
      } else if (state === 12) {
        if (ch === 0x2e) {
          this.state = 29;
        } else {
          this.state = REJECT;
        }
      } else if (state === 13) {
        if (ch === 0x2e) {
          this.state = 30;
        } else if (ch === 0x45 || ch === 0x65) {
          this.state = 31;
        } else {
          this.state = START;
          return this.emit('NUMBER', this.index, input);
        }
      } else if (state === 14) {
        while (ch >= 0x30 && ch <= 0x39) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch === 0x2e) {
          this.state = 30;
        } else if (ch === 0x45 || ch === 0x65) {
          this.state = 31;
        } else {
          this.state = START;
          return this.emit('NUMBER', this.index, input);
        }
      } else if (state === 18) {
        while (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x7a) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        this.state = START;
        return this.emit('NAME', this.index, input);
      } else if (state === 21) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x71 || ch >= 0x73 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x72) {
          this.state = 32;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 22) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x6d || ch >= 0x6f && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x6e) {
          this.state = 33;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 26) {
        while (ch === 0x09 || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch === 0x5c) {
          this.state = 28;
        } else if (ch === 0x22) {
          this.state = START;
          return this.emit('STRING_VALUE', this.index + 1, input);
        } else {
          this.state = REJECT;
        }
      } else if (state === 27) {
        if (ch === 0x22) {
          this.state = 35;
        } else {
          this.state = START;
          return this.emit('STRING_VALUE', this.index, input);
        }
      } else if (state === 28) {
        if (ch === 0x22 || ch === 0x2f || ch === 0x62 || ch === 0x66 || ch === 0x6e || ch === 0x72 || ch === 0x74) {
          this.state = 26;
        } else if (ch === 0x5c) {
          this.state = 36;
        } else if (ch === 0x75) {
          this.state = 37;
        } else {
          this.state = REJECT;
        }
      } else if (state === 29) {
        if (ch === 0x2e) {
          this.state = START;
          return this.emit('ELLIPSIS', this.index + 1, input);
        } else {
          this.state = REJECT;
        }
      } else if (state === 30) {
        if (ch >= 0x30 && ch <= 0x39) {
          this.state = 39;
        } else {
          this.state = REJECT;
        }
      } else if (state === 31) {
        if (ch === 0x2b || ch === 0x2d) {
          this.state = 40;
        } else if (ch >= 0x30 && ch <= 0x39) {
          this.state = 41;
        } else {
          this.state = REJECT;
        }
      } else if (state === 32) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x62 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x61) {
          this.state = 42;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 33) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x7a) {
          this.state = 18;
        } else {
          this.state = START;
          return this.emit('ON', this.index, input);
        }
      } else if (state === 35) {
        while (ch >= 0x09 && ch <= 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch === 0x22) {
          this.state = 43;
        } else if (ch === 0x5c) {
          this.state = 44;
        } else {
          this.state = REJECT;
        }
      } else if (state === 36) {
        while (ch === 0x5c) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch === 0x2f || ch === 0x62 || ch === 0x66 || ch === 0x6e || ch === 0x72 || ch === 0x74) {
          this.state = 26;
        } else if (ch === 0x75) {
          this.state = 37;
        } else {
          this.state = REJECT;
        }
      } else if (state === 37) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x46 || ch >= 0x61 && ch <= 0x66) {
          this.state = 45;
        } else {
          this.state = REJECT;
        }
      } else if (state === 39) {
        while (ch >= 0x30 && ch <= 0x39) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch === 0x45 || ch === 0x65) {
          this.state = 31;
        } else {
          this.state = START;
          return this.emit('NUMBER', this.index, input);
        }
      } else if (state === 40) {
        if (ch >= 0x30 && ch <= 0x39) {
          this.state = 41;
        } else {
          this.state = REJECT;
        }
      } else if (state === 41) {
        while (ch >= 0x30 && ch <= 0x39) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        this.state = START;
        return this.emit('NUMBER', this.index, input);
      } else if (state === 42) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x66 || ch >= 0x68 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x67) {
          this.state = 46;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 43) {
        if (ch >= 0x09 && ch <= 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 35;
        } else if (ch === 0x5c) {
          this.state = 44;
        } else if (ch === 0x22) {
          this.state = 47;
        } else {
          this.state = REJECT;
        }
      } else if (state === 44) {
        while (ch === 0x5c) {
          this.index++;
          ch = this.index < length ? input.charCodeAt(this.index) : -1;
        }
        if (ch >= 0x09 && ch <= 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 35;
        } else if (ch === 0x22) {
          this.state = 48;
        } else {
          this.state = REJECT;
        }
      } else if (state === 45) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x46 || ch >= 0x61 && ch <= 0x66) {
          this.state = 49;
        } else {
          this.state = REJECT;
        }
      } else if (state === 46) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x6c || ch >= 0x6e && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x6d) {
          this.state = 50;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 47) {
        if (ch >= 0x09 && ch <= 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 35;
        } else if (ch === 0x5c) {
          this.state = 44;
        } else if (ch === 0x22) {
          this.state = START;
          return this.emit('BLOCK_STRING_VALUE', this.index + 1, input);
        } else {
          this.state = REJECT;
        }
      } else if (state === 48) {
        if (ch >= 0x09 && ch <= 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x21 || ch >= 0x23 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 35;
        } else if (ch === 0x5c) {
          this.state = 44;
        } else if (ch === 0x22) {
          this.state = 52;
        } else {
          this.state = REJECT;
        }
      } else if (state === 49) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x46 || ch >= 0x61 && ch <= 0x66) {
          this.state = 53;
        } else {
          this.state = REJECT;
        }
      } else if (state === 50) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x64 || ch >= 0x66 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x65) {
          this.state = 54;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 52) {
        if (ch === 0x0a || ch === 0x0d || ch >= 0x20 && ch <= 0x5b || ch >= 0x5d && ch <= 0xffff) {
          this.state = 35;
        } else if (ch === 0x5c) {
          this.state = 44;
        } else {
          this.state = REJECT;
        }
      } else if (state === 53) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x46 || ch >= 0x61 && ch <= 0x66) {
          this.state = 26;
        } else {
          this.state = REJECT;
        }
      } else if (state === 54) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x6d || ch >= 0x6f && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x6e) {
          this.state = 55;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 55) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x73 || ch >= 0x75 && ch <= 0x7a) {
          this.state = 18;
        } else if (ch === 0x74) {
          this.state = 56;
        } else {
          this.state = START;
          return this.emit('NAME', this.index, input);
        }
      } else if (state === 56) {
        if (ch >= 0x30 && ch <= 0x39 || ch >= 0x41 && ch <= 0x5a || ch === 0x5f || ch >= 0x61 && ch <= 0x7a) {
          this.state = 18;
        } else {
          this.state = START;
          return this.emit('FRAGMENT', this.index, input);
        }
      } else if (state === REJECT) {
        throw new Error('Failed to recognize token');
      } else {
        throw new Error('Unexpected state');
      }
      this.index++;
    }
    return null;
  }
}
/**
 * @param {string} input
 * @returns {Generator<Token, void, unknown>}
 */
export default function *lex(input: string) {
  const lexer = new Lexer(input);
  while (true) {
    const token = lexer.next();
    if (token === null) {
      return;
    } else {
      yield token;
    }
  }
}
"
`;
